### PR TITLE
refactor crosslink code out of cross-shard-tx; add more strict verifi…

### PR DESCRIPTION
**Pull request by chaosma**

### CODE REVIEW

**Comment by rlan35** [comment]
Please mark the part you changed in comments.

File: node/node_cross_link.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+	mask, err := bls_cosi.NewMask(committerKeys, nil)
+	if err != nil {
+		return ctxerror.New("[VerifyCrossLink] cannot create group sig mask", "shardID", cl.ShardID(), "blockNum", cl.BlockNum()).WithCause(err)
+	}
+	if err := mask.SetMask(cl.Bitmap()); err != nil {
+		return ctxerror.New("[VerifyCrossLink] cannot set group sig mask bits", "shardID", cl.ShardID(), "blockNum", cl.BlockNum()).WithCause(err)
+	}
+
+	decider := quorum.NewDecider(quorum.SuperMajorityStake)
```
Comment (fxfactorial) [inline node/node_cross_link.go]: do need the shardID Provider: like: 

```go
d := quorum.NewDecider(quorum.SuperMajorityStake)
		d.SetShardIDProvider(func() (uint32, error) {
			return header.ShardID(), nil
		})
		d.SetMyPublicKeyProvider(func() (*bls.PublicKey, error) {
			return nil, nil
		})
```

File: node/node_cross_link.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	"github.com/harmony-one/harmony/consensus/quorum"
+	"github.com/harmony-one/harmony/core/types"
+	bls_cosi "github.com/harmony-one/harmony/crypto/bls"
+	"github.com/harmony-one/harmony/internal/ctxerror"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/harmony-one/harmony/shard"
+)
+
+const (
+	maxPendingCrossLinkSize = 300
```
Comment (chaosma) [inline node/node_cross_link.go]: add maximum size limit

File: node/node_cross_link.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+		}
+	}
+	return nil
+}
+
+// ProcessCrossLinkMessage verify and process Node/CrossLink message into crosslink when it's valid
+func (node *Node) ProcessCrossLinkMessage(msgPayload []byte) {
+	if node.NodeConfig.ShardID == 0 {
+		pendingCLs, err := node.Blockchain().ReadPendingCrossLinks()
+		if err == nil && len(pendingCLs) >= maxPendingCrossLinkSize {
```
Comment (chaosma) [inline node/node_cross_link.go]: limit maximum size

File: node/node_cross_link.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+				Msg("[ProcessingCrossLink] Crosslink Message Broadcast Unable to Decode")
+			return
+		}
+
+		candidates := []types.CrossLink{}
+		utils.Logger().Debug().
+			Msgf("[ProcessingCrossLink] Received crosslinks: %d", len(crosslinks))
+
+		for i, cl := range crosslinks {
+			if i > crossLinkBatchSize {
```
Comment (chaosma) [inline node/node_cross_link.go]: limit incoming size

File: node/node_cross_link.go (hunk 1 above)
Comment (chaosma) [inline node/node_cross_link.go]: add voting power checks for crosslink

**Pull request merged**